src/main/java/com/sarality/db/InMemoryTable.java

```diff
@@ -41,8 +42,22 @@ public Long create(T data) {
 
   @Override
   public List<T> readAll(Query query) {
+
     List<T> dataList = new ArrayList<>();
-    dataList.addAll(dataMap.values());
+
+    if (query == null) {
+      dataList.addAll(dataMap.values());
+      return dataList;
+    }
+
+    String[] whereArgs = query.getWhereClauseArguments();
+
+    if (whereArgs.length != 0) {
```

Comment:
Make this a length == 1 check and throw an Unsupported Exception or IllegalArgument Exception when it is not the case so that it doesn't fail silently.

Add a TODO to make sure that the where clause starts with Column name for primary key or it of the format "ProductId = ?". If you can implement this, even better. Again, throw an exception when that's not the case.